lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -734,46 +783,189 @@ void LowerLayersPass::preprocessLayers(CircuitNamespace &ns, LayerOp layer,
+    // We will also have to manually enable any subsequent ancestors.
+    auto parentMacroSymbol = macroNames[parent];
+    b.create<sv::MacroDefOp>(loc, parentMacroSymbol,
+                             StringAttr::get(&getContext()));
+    parent = layer->getParentOfType<LayerOp>();
+  }
+
+  // Create IR to include bind files for child modules.
+  for (auto *record : *node) {
+    auto *child = record->getTarget()->getModule().getOperation();
+    auto files = bindFiles[child];
+    auto lookup = files.find(layer);
+    if (lookup != files.end())
+      b.create<sv::IncludeOp>(loc, IncludeStyle::Local, lookup->second.name);
+  }
+
+  // Save the bind file information for later.
+  auto &info = bindFiles[op][layer];
+  info.name = bindFileNameAttr;
+  info.body = guard.getElseBlock();
+}
+
+void LowerLayersPass::preprocessModule(CircuitNamespace &ns,
+                                       InstanceGraphNode *node) {
+  auto *op = node->getModule().getOperation();
+  if (!op)
+    return;
+
+  auto module = dyn_cast<FModuleOp>(op);
+  if (!module)
+    return;
+
+  OpBuilder b(&getContext());
+  b.setInsertionPointAfter(module);
+
+  // Create a bind file only if the layer is used under the module.
+  llvm::SmallDenseSet<LayerOp> todo;
+
+  // If the module is public, create a bind file for all layers.
+  if (module.isPublic())
+    for (auto [_, layer] : symbolToLayer)
+      if (layer.getConvention() == LayerConvention::Bind)
+        todo.insert(layer);
+
+  // Handle layers used directly in this module.
+  module->walk([&](LayerBlockOp layerBlock) {
+    auto layer = symbolToLayer[layerBlock.getLayerNameAttr()];
+    if (layer.getConvention() == LayerConvention::Inline)
+      return;
+
+    // Create a bindfile for any layer directly used in the module.
+    todo.insert(layer);
+
+    // Determine names for all modules that will be created.
+    auto moduleName = module.getModuleName();
+    auto layerName = layerBlock.getLayerName();
+
+    // A name hint for the module created from this layerblock.
+    auto layerBlockModuleName = moduleNameForLayer(moduleName, layerName);
+
+    // A name hint for the hier-path-op which targets the bound-in instance of
+    // the module created from this layerblock.
+    auto layerBlockHierPathName = hierPathNameForLayer(moduleName, layerName);
+
+    LayerBlockGlobals globals;
+    globals.moduleName = ns.newName(layerBlockModuleName);
+    globals.hierPathName = ns.newName(layerBlockHierPathName);
+    layerBlockGlobals.insert({layerBlock, globals});
+  });
+
+  // Create a bindfile for layers used indirectly under this module.
+  for (auto *record : *node) {
+    auto *child = record->getTarget()->getModule().getOperation();
+    for (auto [layer, _] : bindFiles[child])
+      todo.insert(layer);
+  }
+
+  // Build the bindfiles for any layer seen under this module.
+  for (auto [sym, layer] : symbolToLayer)
+    if (todo.contains(layer))
+      buildBindFile(ns, node, b, sym, layer);
+}
```

Comment:
This could use a `SetVector` to avoid iteration over all layers seeing if we get a hit. However, this would make the bind file order be layer block order instead of layer order (unless insertion order was even more complicated).

This is probably fine given that the number of layers is relatively small. If this changes, we may need to revisit this.

Reply:
Assuming the intention is to keep the bind files in layer declaration order, can you include a comment to this effect?